### PR TITLE
fix issues affecting code quality

### PR DESCRIPTION
**Pull request by withshubh**
### Description
Hi :wave: 
I ran the [DeepSource](https://deepsource.io) analysis on the forked copy of this repo and found some interesting [code quality issues](https://deepsource.io/gh/withshubh/pytest/issues/?category=recommended).

### Motivation and Context
Some of the changes boost minor performance and a few removes the anti-patterns in code.

### Summary of changes

- Removed length check in favour of truthiness of the object
- Used literal syntax instead of function calls to create data structure
- Refactored useless `else` block in the loop
- Replaced ternary syntax with if expression
- Used `callable()` to check if the object is calllable
- Used `sys.exit()` calls
- Added .deepsource.toml to continuously analyze repo for code quality issues

### CODE REVIEW

File: src/_pytest/pathlib.py [hunk 1]
```diff
@@ -226,11 +226,11 @@ def make_numbered_dir(root: Path, prefix: str) -> Path:
         else:
             _force_symlink(root, prefix + "current", new_path)
             return new_path
-    else:
-        raise OSError(
-            "could not create numbered dir with prefix "
-            "{prefix} in {root} after 10 tries".format(prefix=prefix, root=root)
-        )
+
```
Comment (RonnyPfannschmidt) [inline src/_pytest/pathlib.py:229]: this one seems potentially problematic

**Review by RonnyPfannschmidt — APPROVED** [review]
i like the changes, many of those idioms are leftovers from past times where python3 or python2 had different needs

File: src/_pytest/_code/code.py [hunk 2]
```diff
@@ -1221,7 +1221,7 @@ def getfslineno(obj: object) -> Tuple[Union[str, Path], int]:
         except TypeError:
             return "", -1
 
-        fspath = fn and absolutepath(fn) or ""
```
Comment (RonnyPfannschmidt) [inline src/_pytest/_code/code.py:1224]: it seems that this one confuses mypy and has to be typed as `Union[str, Path]`

**Comment by withshubh** [comment]
Hey @RonnyPfannschmidt :wave: 
I have revert those fixes. Thanks for the review :sparkling_heart:

File: src/_pytest/pathlib.py (hunk 1 above)
Comment (The-Compiler) [reply src/_pytest/pathlib.py]: I think it's fine, no? The for-loop will always run (`range(10)` is never empty), and in the for-loop, either we succeed and `return`, or we fail 10 times and land here.

File: src/_pytest/_code/code.py (hunk 2 above)
Comment (The-Compiler) [reply src/_pytest/_code/code.py]: @withshubh I think it'd still be good to have this, can you try whether `fspath: Union[str, Path] = absolutepath(fn) if fn else ""` makes things work?

File: .deepsource.toml [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+    "changelog/**",
+    "doc/en/**"
+]
+
+[[analyzers]]
+name = "python"
+enabled = true
+
+  [analyzers.meta]
+  runtime_version = "3.x.x"
```
Comment (asottile) [inline .deepsource.toml:15]: we're not using this vendor so I don't think we should add this file

from my previous experience with deepsource it is more of an annoyance than a benefit -- in particular there is no way to run it locally

**Review by asottile — CHANGES_REQUESTED** [review]
in case it's not clear, my comment is blocking

**Comment by withshubh** [comment]
> in case it's not clear, my comment is blocking

I totally understand @asottile :)

**Comment by nicoddemus** [comment]
Thanks folks for the PR/reviews.

Closing for now then.

**Pull request closed without merging**